File: Languages/cpp/algorithms-dsa/leetcode/sliding-window/longest_substring_without_repeating.cpp
/**
 * LeetCode 3: Longest Substring Without Repeating Characters
 * Topic: Sliding Window
 * Difficulty: Medium
 *
 * Given a string s, find the length of the longest substring without repeating characters.
 *
 * Approach: Sliding window with a hashset. Expand the right pointer, if a duplicate is
 * found shrink from the left until the duplicate is removed. Track the max window size.
 *
 * Time: O(n), Space: O(min(n, alphabet))
 */
#include <iostream>
#include <string>
#include <unordered_set>
#include <algorithm>
#include <cassert>
using namespace std;

int lengthOfLongestSubstring(string s) {
    unordered_set<char> charSet;
    int left = 0;
    int maxLen = 0;

    for (int right = 0; right < (int)s.size(); right++) {
        while (charSet.count(s[right])) {
            charSet.erase(s[left]);
            left++;
        }
        charSet.insert(s[right]);
        maxLen = max(maxLen, right - left + 1);
    }

    return maxLen;
}

int main() {
    assert(lengthOfLongestSubstring("abcabcbb") == 3);
    assert(lengthOfLongestSubstring("bbbbb") == 1);
    assert(lengthOfLongestSubstring("pwwkew") == 3);
    assert(lengthOfLongestSubstring("") == 0);
    assert(lengthOfLongestSubstring(" ") == 1);
    assert(lengthOfLongestSubstring("au") == 2);
    assert(lengthOfLongestSubstring("dvdf") == 3);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/python/algorithms-dsa/leetcode/sliding-window/longest_substring_without_repeating.py
"""
LeetCode 3: Longest Substring Without Repeating Characters
Topic: Sliding Window
Difficulty: Medium

Given a string s, find the length of the longest substring without repeating characters.

Approach: Sliding window with a hashset. Expand the right pointer, if a duplicate is
found shrink from the left until the duplicate is removed. Track the max window size.

Time: O(n), Space: O(min(n, alphabet))
"""


def length_of_longest_substring(s: str) -> int:
    char_set: set[str] = set()
    left = 0
    max_len = 0

    for right in range(len(s)):
        while s[right] in char_set:
            char_set.remove(s[left])
            left += 1
        char_set.add(s[right])
        max_len = max(max_len, right - left + 1)

    return max_len


if __name__ == "__main__":
    assert length_of_longest_substring("abcabcbb") == 3
    assert length_of_longest_substring("bbbbb") == 1
    assert length_of_longest_substring("pwwkew") == 3
    assert length_of_longest_substring("") == 0
    assert length_of_longest_substring(" ") == 1
    assert length_of_longest_substring("au") == 2
    assert length_of_longest_substring("dvdf") == 3

    print("All tests passed!")
